### PR TITLE
Deduplicate some code and compile-time values around vtables

### PR DESCRIPTION
**Pull request by oli-obk**
r? @RalfJung

### CODE REVIEW

**Comment by oli-obk** [comment]
ups, ordering issues

@bors try

**Comment by oli-obk** [comment]
@rust-timer build fce7dc2549f6c3665405a744e79bfa61868e2b0a

**Comment by Mark-Simulacrum** [comment]
@rust-timer build fce7dc2549f6c3665405a744e79bfa61868e2b0a

Bot should now recognize @oli-obk .

**Comment by Mark-Simulacrum** [comment]
@rust-timer build 178f3ba362979973aef288e95b6730595586ce00

**Comment by oli-obk** [comment]
https://perf.rust-lang.org/compare.html?start=e9e27e6a6258b3adf00a5dd35d2676656224880d&end=178f3ba362979973aef288e95b6730595586ce00&stat=max-rss undoes the memory regression introduced in https://github.com/rust-lang/rust/pull/54461#issuecomment-429282619 but seems to regress load of others?!

Ok I'm doing more in this PR than just undoing the regression, I'll split out the changes of this PR into separate PRs so we can observe them in isolation

File: src/librustc_mir/interpret/eval_context.rs [hunk 1]
```diff
@@ -50,6 +51,9 @@ pub struct EvalContext<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'a, 'mir, 'tcx>> {
 
     /// The virtual call stack.
     pub(crate) stack: Vec<Frame<'mir, 'tcx, M::PointerTag>>,
+
+    /// A cache for deduplicating vtables
+    pub(super) vtables: FxHashMap<(Ty<'tcx>, ty::PolyTraitRef<'tcx>), AllocId>,
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs]: I am a bit surprised to see this here and not in `Memory`, but I guess it does not matter much.

File: src/librustc_mir/interpret/eval_context.rs [hunk 2]
```diff
@@ -207,6 +211,7 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tc
             param_env,
             memory: Memory::new(tcx, memory_data),
             stack: Vec::new(),
+            vtables: FxHashMap(),
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs]: I think we are supposed to use `FxHashMap::default()` these days.

File: src/librustc_mir/interpret/traits.rs [hunk 3]
```diff
@@ -24,24 +24,32 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
     pub fn get_vtable(
         &mut self,
         ty: Ty<'tcx>,
-        trait_ref: ty::PolyTraitRef<'tcx>,
+        poly_trait_ref: Option<ty::PolyExistentialTraitRef<'tcx>>,
     ) -> EvalResult<'tcx, Pointer<M::PointerTag>> {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/traits.rs]: Please add a comment here saying which other method this has to be kept in sync with, and add a matching comment on the other end.

Comment (RalfJung) [reply src/librustc_mir/interpret/traits.rs]: Oh never mind, you did.  That's the issue with reviewing commits individually.^^

**Comment by RalfJung** [comment]
r=me with the lints fixed and perf happy.

**Comment by RalfJung** [comment]
@rust-timer build b4b95a37ef53d88772feefb0d5100a1774915ed3

**Comment by oli-obk** [comment]
So my changes now are not causing perf regressions anymore (actually improving perf on some tests), but there are still some peak memory usage regressions.

Most notably

* webrender-debug baseline incremental (6%)
    * not sure how strongly this fluctuates, the previous run had 3%
* coercions-debug baseline incremental (3%, marked with ? for fluctuations)
* ripgrep-debug clean (3%, no regression in the previous run)
* inflate-debug patched incremental: println (2%)

There's also a big swath of small (1%) improvements

File: src/librustc_mir/interpret/eval_context.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: done

**Comment by RalfJung** [comment]
Time, guessing spurious.

@bors retry

**Pull request merged**